src/Hamiltonians/correlation_functions.jl

```diff
@@ -342,25 +342,57 @@ function diagonal_element(::StringCorrelator{0}, add::SingleComponentFockAddress
     return result / M
 end
 
+# function diagonal_element(::StringCorrelator{D}, add::SingleComponentFockAddress) where {D}
+#     M = num_modes(add)
+#     N = num_particles(add)
+#     n̄ = N/M
+#     d = mod(D, M)
+#     v = onr(add)
+
+#     result = 0  
+#     for i in eachindex(v)
+#         phase_sum = sum( (v[mod1(k,M)] - n̄) for k in i:1:(i+d-1) )
+
+#         result += (v[i]- n̄) * exp(pi * im * phase_sum) * (v[mod1(i + d, M)]-n̄)
+#     end
+
+#     if M == N
+#         return real(result) /M
+#     end
+
+#     return result / M
+# end
+
+num_offdiagonals(::StringCorrelator, ::SingleComponentFockAddress) = 0
+
 function diagonal_element(::StringCorrelator{D}, add::SingleComponentFockAddress) where {D}
     M = num_modes(add)
     N = num_particles(add)
-    n̄ = N/M
     d = mod(D, M)
     v = onr(add)
 
-    result = 0  
-    for i in eachindex(v)
-        phase_sum = sum( (v[mod1(k,M)] - n̄) for k in i:1:(i+d-1) )
+    if iszero(N % M)
+        n̄ = N ÷ M
+        
+        result = 0  
+        for i in eachindex(v)
+            phase_sum = sum( (v[mod1(k,M)] - n̄) for k in i:1:(i+d-1) )
 
-        result += (v[i]- n̄) * exp(pi * im * phase_sum) * (v[mod1(i + d, M)]-n̄)
-    end
-
-    if M == N
-        return real(result) /M
-    end
+            result += (v[i]- n̄) * exp(pi * im * phase_sum) * (v[mod1(i + d, M)]-n̄)
```

Comment:
You are in the branch where `phase_sum` should be an integer and you have branched in order to avoid complex numbers. Have you tested and verified this function to return a real-valued number type?

I believe that you'd have to write it like this if you want to avoid complex numbers:
```suggestion
            result += (v[i]- n̄) * (-1)^phase_sum * (v[mod1(i + d, M)]-n̄)
```

Reply:
The code on line 381 on my local version looks like the correction suggestion that you made, not the red one. I think I changed this on a commit later but never pushed this commit

Reply:
Please push it then, such that we can see the latest version of the code.